UnrealPlugin/Source/PlayfabGSDK/Private/FGSDKInternal.cpp

```diff
@@ -0,0 +1,374 @@
+﻿#include "FGSDKInternal.h"
+
+#include "GSDKConfiguration.h"
+#include "GSDKUtils.h"
+#include "HttpModule.h"
+#include "Async/Async.h"
+#include "Json.h"
+#include "PlayfabGSDK.h"
+#include "Interfaces/IHttpResponse.h"
+#include "Misc/Paths.h"
+#include "Logging/LogMacros.h"
+#include "HAL/Event.h"
+
+FGSDKInternal::FGSDKInternal()
+	: TransitionToActiveEvent(FPlatformProcess::GetSynchEventFromPool(false))
+	, SignalHeartbeatEvent(FPlatformProcess::GetSynchEventFromPool(false))
+{
+	TUniquePtr<FConfiguration> ConfigPtr;
+
+	FString FileName = UGSDKUtils::GetEnvironmentVariable(TEXT("GSDK_CONFIG_FILE"));
+
+	// If the configuration file is not there, we'll get our config from environment variables
+	if (!FPaths::FileExists(FileName))
+	{
+		ConfigPtr = MakeUnique<FEnvironmentVariableConfiguration>();
+	}
+	else
+	{
+		ConfigPtr = MakeUnique<FJsonFileConfiguration>(FileName);
+	}
+
+	TMap<FString, FString> GameCerts = ConfigPtr->GetGameCertificates();
+	for (const auto& GameCert: GameCerts)
+	{
+		ConfigSettings.Add(GameCert);
+	}
+
+	TMap<FString, FString> MetaDatas = ConfigPtr->GetBuildMetadata();
+	for (const auto& MetaData: MetaDatas)
+	{
+		ConfigSettings.Add(MetaData);
+	}
+
+	TMap<FString, FString> Ports = ConfigPtr->GetGamePorts();
+	for (const auto& Port: Ports)
+	{
+		ConfigSettings.Add(Port);
+	}
+
+	ConfigSettings.Add(FPlayfabGSDKModule::HEARTBEAT_ENDPOINT_KEY, ConfigPtr->GetHeartbeatEndpoint());
+	ConfigSettings.Add(FPlayfabGSDKModule::SERVER_ID_KEY, ConfigPtr->GetServerId());
+	ConfigSettings.Add(FPlayfabGSDKModule::LOG_FOLDER_KEY, ConfigPtr->GetLogFolder());
+	ConfigSettings.Add(FPlayfabGSDKModule::SHARED_CONTENT_FOLDER_KEY, ConfigPtr->GetSharedContentFolder());
+	ConfigSettings.Add(FPlayfabGSDKModule::CERTIFICATE_FOLDER_KEY, ConfigPtr->GetCertificateFolder());
+	ConfigSettings.Add(FPlayfabGSDKModule::TITLE_ID_KEY, ConfigPtr->GetTitleId());
+	ConfigSettings.Add(FPlayfabGSDKModule::BUILD_ID_KEY, ConfigPtr->GetBuildId());
+	ConfigSettings.Add(FPlayfabGSDKModule::REGION_KEY, ConfigPtr->GetRegion());
+	ConfigSettings.Add(FPlayfabGSDKModule::PUBLIC_IP_V4_ADDRESS_KEY, ConfigPtr->GetPublicIpV4Address());
+	ConfigSettings.Add(FPlayfabGSDKModule::FULLY_QUALIFIED_DOMAIN_NAME_KEY, ConfigPtr->GetFullyQualifiedDomainName());
+
+	if (ConfigSettings[FPlayfabGSDKModule::HEARTBEAT_ENDPOINT_KEY].IsEmpty() || ConfigSettings[FPlayfabGSDKModule::SERVER_ID_KEY].IsEmpty())
+	{
+		UE_LOG(LogPlayfabGSDK, Fatal, TEXT("Heartbeat endpoint and Server id are required configuration values."));
+		return;
+	}
+
+	ConnectionInfo = ConfigPtr->GetGameServerConnectionInfo();
+
+	if (ConfigPtr->ShouldLog())
+	{
+		StartLog();
+	}
+
+	// Use highest frequency permitted heartbeat interval until VMAgent tells an updated one.
+	NextHeartbeatIntervalMs = 1000; // TODO Move to constant
```

Comment:
Suggestion:
ConfigPtr->GetDefaultHeartbeatInterval();

Reply:
If I add this, it would be a constant defined in the config object, as the json config does not contain a value for this yet, but I can add it.